File: Harbor.Domain/Pages/PageComponents/PageLink.cs
using System;
using System.Collections.Generic;
using Harbor.Domain.Pages.PageResources;

namespace Harbor.Domain.Pages.PageComponents
{
	public class PageLink : PageComponent
	{
		public PageLink(Page page, string uicid) : base(page, uicid)
		{
			if (IsNew() == false)
			{
				_page = page.GetPageLink(PageID);
			}
		}

		private Page _page;

		public bool IsNew()
		{
			return PageID == 0;
		}

		public string Title
		{
			get
			{
				return _page == null ? null : _page.Title;
			}
		}
		
		public int PageID
		{
			get
			{
				var id = GetProperty("pageID");
				return id == null ? 0 : int.Parse(id);
			}
		}

		public string PreviewText
		{
			get
			{
				return _page == null ? null : _page.PreviewText;
			}
		}

		public Guid? PreviewImageID
		{
			get
			{
				return _page == null ? null : _page.PreviewImageID;
			}
		}

		public bool HasPreviewImage
		{
			get
			{
				return PreviewImageID != null;
			}
		}

		public string TileDisplay
		{
			get 
			{
				var display = GetProperty("tileDisplay");
				return display ?? "normal";
			}
		}

		public string VirtualPath
		{
			get
			{
				return _page == null ? null : _page.VirtualPath;
			}
		}

		public override IEnumerable<PageResource> DeclareResources()
		{
			if (PageID == 0)
			{
				yield break;
			}

			yield return new PageLinkResource(Page, pageID: PageID);
		}
	}
}

File: Harbor.UI/Scripts/apps/PageEditor/comps/pageLink/pageLink.js

var pageLink = module("pageLink").use("pageComponent", "bbext");


pageLink.component("pageLink", function (viewFactory) {

	this.view = viewFactory.create("pageLinkView", {
		el: this.$el,
		model: this.model,
		uicid: this.uicid
	});

}, {
	$inject: ["viewFactory"],
	
    modelType: "pageLinkModel",

    initialize: function () {
        
    },

	create: function () {
	    this.open();
	    this.view.openPageSelector();
	},
	
	open: function () {
		this.view.render();
	},
	
	close: function () {
		this.view.close();
	}
});

pageLink.model("pageLinkModel", {
	component:  {
		pageProperties: ["pageID", "tileDisplay"],
	
		getDefaults: function (page, pageProperties) {
			return _.pick(page.getPageLink(pageProperties.pageID),
				"title", "previewText", "previewImageID", "link");
		}
	},
	
	defaults: {
		pageID: 0,
		tileDisplay: "normal",
		title: null,
		previewText: null,
		previewImageID: null,
		previewImageSrc: null,
		tileClassName: "tile",
		link: null
	},
	
	hasPageLink: function () {
		return this.get("pageID") > 0 ? true : false;
	},
	
	previewImageSrc: {
		get: function (value) {
			var src = Application.url("file/" +
				this.get("previewImageID") + "/preview.img?res=low");
			return src;
		}
	},
	
	tileClassName: {
		get: function (value) {
			var display = this.get("tileDisplay");
			return display === "wide" ? "tile tile-wide" : "tile";
		}
	}
});


pageLink.view("pageLinkView", {
	initialize: function () {
		this.listenTo(this.model, "change:tileDisplay", this.save);
	},
	
	events: {
		"click [data-rel=select]": function () {
			this.openPageSelector();
		},
		"click a": function (event) {
			// prevent link navigation while editing
			event.preventDefault();
		}
	},
	
	render: function () {
		var editDiv =  this.template("PageLink-Edit")();
		var imgCtr = this.$("[data-rel=pageLink]");
		imgCtr.after(editDiv);
		this.bindModelToView();
	},
	
	renderPageLink: function () {
		if (this.model.hasPageLink() === false) {
			return;
		}
		this.renderTemplate("PageLink")(this.model.toJSON());
		this.render();
	},
	
	save: function () {
		AjaxRequest.handle(this.model.save(), {
			success: this.renderPageLink
		}, this);
	},

	openPageSelector: function () {
		PageSettings.events.trigger("modal:opened");
		PageSelector.start({
			region: PageEditor.regions.modal,
			close: function () {
				PageSettings.events.trigger("modal:closed");
			},
			select: function (selectedPage) {
				var pageID = selectedPage.get("id");
				this.model.set({
					pageID: pageID,
					title: selectedPage.get("title"),
					previewText: selectedPage.get("previewText"),
					previewImageID: selectedPage.get("previewImageID")
				});
				this.save();
			}
		}, this);
	},
	
	onClose: function () {
		this.$("[data-rel=edit]").remove();
	}
});